### PR TITLE
Feature dev server task

### PR DESCRIPTION
**Pull request by dgreene-r7**
This PR creates self-contained S3 and Metadata servers to mock dependencies and allow local development. Simply run `npm run dev-server` and you should be all set.

### CODE REVIEW

File: test/bin/s3-server.js [hunk 1]
```diff
@@ -73,55 +83,107 @@ function createTempDirForBucket() {
...
+/**
+ * Copies files from the watched folder the the "bucket"
+ * @param {String} bucket
+ */
+function copyFiles(bucket) {
+  const walker = walk.walk(path);
+
+  walker.on('file', (root, fileStats, next) => {
+    const pathToFile = `${Path.relative(path, root)}/${fileStats.name}`;
+    const stream = fs.createReadStream(`${path}/${pathToFile}`);
```
Comment (fmitchell-r7) [inline test/bin/s3-server.js]: Any reason not to use `Path.join` instead of string interpolation?

File: test/bin/s3-server.js [hunk 2]
```diff
@@ -73,55 +83,107 @@ function createTempDirForBucket() {
...
+    });
   });
+}
 
-  client.run((serverErr, host, p) => {
-    if (serverErr) {
+function init() {
+  fs.stat(path, (err, stats) => {
+    if (err || !stats.isDirectory()) {
       process.exit(1);
```
Comment (fmitchell-r7) [inline test/bin/s3-server.js]: Does it make sense to log something here before exiting so the user knows why the process is exiting?

File: test/bin/s3-server.js [hunk 3]
```diff
@@ -73,55 +83,107 @@ function createTempDirForBucket() {
...
+      silent: false,
+      directory: tmpDir
+    });
 
-    awsClient.createBucket({Bucket: bucket}, (createBucketErr) => {
-      if (createBucketErr) {
-        throw createBucketErr;
+    client.run((serverErr, host, p) => {
+      if (serverErr) {
+        process.exit(1);
```
Comment (fmitchell-r7) [inline test/bin/s3-server.js]: Does it make sense to log something here before exiting so the user knows why the server is exiting?

File: test/bin/s3-server.js (hunk 1 above)
Comment (dgreene-r7) [inline test/bin/s3-server.js]: Nope, not in particular. Brainfarted and forgot that it existed.

File: test/bin/s3-server.js [hunk 4]
```diff
@@ -73,55 +83,113 @@ function createTempDirForBucket() {
...
-    process.exit(1);
+/**
+ * Copies files from the watched folder the the "bucket"
+ * @param {String} bucket
+ */
+function copyFiles(bucket) {
+  const walker = walk.walk(path);
+
+  walker.on('file', (root, fileStats, next) => {
+    const pathToFile = `${Path.relative(path, root)}/${fileStats.name}`;
```
Comment (fmitchell-r7) [inline test/bin/s3-server.js]: Can you use `Path.join` here as well?

File: test/bin/s3-server.js [hunk 5]
```diff
@@ -73,55 +83,113 @@ function createTempDirForBucket() {
...
+    chokidar.watch(path, {persistent: true}).on('unlink', () => onFileChange(bucket));
   });
+}
 
-  client.run((serverErr, host, p) => {
-    if (serverErr) {
+function init() {
+  fs.stat(path, (err, stats) => {
+    if (err || !stats.isDirectory()) {
+      Log.error('Path is not a directory');
```
Comment (fmitchell-r7) [inline test/bin/s3-server.js]: Do we want to log the path as part of the error message so the user knows which one is invalid?

File: test/bin/s3-server.js [hunk 6]
```diff
@@ -73,55 +83,113 @@ function createTempDirForBucket() {
...
+    const pathToFile = `${Path.relative(path, root)}/${fileStats.name}`;
+    const stream = fs.createReadStream(Path.join(path, pathToFile));
+
+    awsClient.putObject({
+      Bucket: bucket,
+      Key: pathToFile,
+      Body: stream
+    }, (err) => {
+      if (err) {
+        Log.info(err, err.stack);
```
Comment (fmitchell-r7) [inline test/bin/s3-server.js]: Should we log this at the `error` level instead of `info`?

File: test/bin/s3-server.js [hunk 7]
```diff
@@ -73,55 +83,113 @@ function createTempDirForBucket() {
...
+      silent: false,
+      directory: tmpDir
+    });
 
-    awsClient.createBucket({Bucket: bucket}, (createBucketErr) => {
-      if (createBucketErr) {
-        throw createBucketErr;
+    client.run((serverErr, host, p) => {
+      if (serverErr) {
+        Log.error(serverErr);
```
Comment (fmitchell-r7) [inline test/bin/s3-server.js]: Do we want to log `serverErr.stack` as well? Would that be useful?

File: test/bin/s3-server.js [hunk 8]
```diff
@@ -73,55 +83,113 @@ function createTempDirForBucket() {
...
+}
+
+/**
+ * Handles cleanup on app exit
+ */
+function exitHandler(err) {
+  let code = 0;
+
+  if (err) {
+    Log.error(err);
```
Comment (fmitchell-r7) [inline test/bin/s3-server.js]: Do we want to log `err.stack` as well? Would that be useful?

File: test/bin/s3-server.js [hunk 9]
```diff
@@ -73,55 +83,113 @@ function createTempDirForBucket() {
...
   }
-  const bucket = getBucketName(path);
-  const tmpDir = createTempDirForBucket();
-
-  const client = new S3rver({
-    port,
-    hostname,
-    silent: false,
-    directory: tmpDir
+  Log.info('Cleaning up temp directory');
```
Comment (fmitchell-r7) [inline test/bin/s3-server.js]: Do we have the temp directory name so we can log that here too? Might be useful for the user to know what directory they need to manually clean up if the clean up process fails.

**Comment by fmitchell-r7** [comment]
:+1: Thanks for getting all these development bits working together. :shipit:

**Pull request merged**